ENTMQBR-1648 Don't hardcode openshift.DNS_PING service name

gaohoward opened the pull request:
Now the service name takes the form of
${APPLICATION_NAME}-${PING_SERVICE_NAME}

If both APPLICATION_NAME and PING_SERVICE_NAME are absent the
DNS_PING service name will be default to 'ping' (that's the
original value). If only APPLICATION_NAME is absent, the service
name is the value of PING_SERVICE_NAME (without prefixed with
the dash).

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [ ] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

RoddieKieley commented:
@gaohoward I grabbed your branch and attempted to create a new project and then process a template to populated it without success:

> 
> 
> [rkieley@i7t450s jboss-amq-7-broker-openshift-image]$ oc process -n amq73-p-c-0 -f templates/amq-broker-73-persistence-clustered.yaml -p APPLICATION_NAME=application3 -p AMQ_USER=admin -p AMQ_PASSWORD=admin -p AMQ_NAME=amq3 -p AMQ_EXTRA_ARGS=--no-autotune | oc create -f -
> serviceaccount "application3-service-account" created
> role "application3-role" created
> rolebinding "application3-role-binding" created
> service "amq3-amq-headless" created
> statefulset "application3-amq" created
> The Service "application3-${PING_SERVICE_NAME}" is invalid: metadata.name: Invalid value: "application3-${PING_SERVICE_NAME}": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')
> [rkieley@i7t450s jboss-amq-7-broker-openshift-image]$ oc delete project/amq73-p-c-0
> project "amq73-p-c-0" deleted
> [rkieley@i7t450s jboss-amq-7-broker-openshift-image]$ oc new-project amq73-p-c-0
> Already on project "amq73-p-c-0" on server "https://master-ocp311:8443".
> 
> You can add applications to this project with the 'new-app' command. For example, try:
> 
>     oc new-app centos/ruby-22-centos7~https://github.com/openshift/ruby-ex.git
> 
> to build a new example application in Ruby.
> [rkieley@i7t450s jboss-amq-7-broker-openshift-image]$ oc process -n amq73-p-c-0 -f templates/amq-broker-73-persistence-clustered.yaml -p APPLICATION_NAME=application3 -p AMQ_USER=admin -p AMQ_PASSWORD=admin -p AMQ_NAME=amq3 -p AMQ_EXTRA_ARGS=--no-autotune -p PING_SERVICE_NAME=ping | oc create -f -
> error: unknown parameter name "PING_SERVICE_NAME"
> error: no objects passed to create
> [rkieley@i7t450s jboss-amq-7-broker-openshift-image]$
> 

What am I missing?

RoddieKieley commented:
@gaohoward Sorry for the delay in getting back to this. I grabbed the latest commit above into a new branch, built a new image and deployed the updated persistent-clustered-ssl template into a new-project and this doesn't work for me. I'm seeing

> 2019-06-27 17:39:51,311 INFO  [org.openshift.ping.common.Utils] 3 attempt(s) with a 1000ms sleep to execute [GetServiceHosts] failed. Last failure was [java.net.UnknownHostException: ping]

when I spin up 2 brokers via

> oc process -n amq74-p-c-ssl-0 -f templates/amq-broker-74-persistence-clustered-ssl.yaml -p APPLICATION_NAME=application0 -p AMQ_USER=admin -p AMQ_PASSWORD=admin -p AMQ_NAME=amq0 -p AMQ_EXTRA_ARGS=--no-autotune -p AMQ_KEYSTORE=broker.ks -p AMQ_TRUSTSTORE=broker.ts -p AMQ_KEYSTORE_PASSWORD=password -p AMQ_TRUSTSTORE_PASSWORD=password | oc create -f -

then

> oc scale statefulset application0-amq --replicas 2

In the logs I also see:

> PING_SERVICE is not set
> APPLICATION_NAME is not set
>  Generating jgroups-ping.xml, current dir is: /home/jboss, AMQHOME: /opt/amq

I would note that the broker instance is in /home/jboss/amq0 in this case while the jgroups_ping.xml appears to be written to the /opt/amq/conf directory where the distro is located.

Also while APPLICATION_NAME and PING_SERVICE_NAME are set as parameters in the template they aren't passed through to the statefulset specification in the [template: spec: containers: env: section](https://github.com/gaohoward/jboss-amq-7-broker-openshift-image/blob/a_73_dev_1648/templates/amq-broker-73-persistence-clustered-ssl.yaml#L268) which means that these aren't available in the containers environment for the broker scripts to use.

As well the [drainer pod template](https://github.com/gaohoward/jboss-amq-7-broker-openshift-image/blob/a_73_dev_1648/templates/amq-broker-73-persistence-clustered-ssl.yaml#L143) will need to be updated as when that spins up a broker it joins the cluster.

Was this ok in a straight kubernetes environment? Wondering if there are some OpenShift specifics here.

gaohoward commented:
Hi @RoddieKieley I'll take a look.

gaohoward commented:
Hi Roddie,

I've retested my PR and compared with yours and so far here is what I've come up with:

1) Regarding this error:

    2019-06-27 17:39:51,311 INFO [org.openshift.ping.common.Utils] 3 attempt(s) with a 1000ms sleep to execute [GetServiceHosts] failed. Last failure was [java.net.UnknownHostException: ping]

It happens on plain kubernetes too. I didn't notice it before because the broker seems started up ok.
However when I scale up the cluster is not formed. I don't know how to fix it yet. However it seems
to be an environment issue. I tried the image without my PR, it also happens.

I also tried deploying the template on quicklab and I saw this service is created:
service/application0-ping created

I'm guessing the broker is failed to access this service. But I haven't figured that out yet.
Somehow on quickLab after I issued the 'oc scale statefulset application0-amq --replicas 2' command
the pod seems get stuck at "ContainerCreating" status. Not sure why.

2) About PING_SERVICE is not set etc:

I've noticed that you are using amq-broker-74-persistence-clustered-ssl.yaml. I assume you are using
74 branch. However my branch is based on amq-broker-73-dev. Can you try amq-broker-73-persistence-clustered-ssl.yaml? This file is updated with the unset parameters. And the jroups-ping.xml is also
updated in my PR and the broker should have no problem picking that up (I can see in my logs that it's loaded).

In general I don't think there is much difference between openshift and kubernetes except that kubernetes may need some extra config to make jgroups-ping work. I'll continue tomorrow and see
If I can find a solution.

RoddieKieley commented:
Hey Howard. I had rebased your branch on the latest 7.4 downstream work in anticipation of merging it, and as there were no meaningful changes to the templates I would have expected it to operator as per 7.3 downstream. When possible I will retest plain vanilla 7.3.

With respect to the application0-ping service, I did see that this was created. However the openshift-dns plugin was still configured to look for a service named 'ping' and thus the cluster would not form as the service was 'unavailable' as 'ping'

gaohoward commented:
Hi Roddie,
Regarding the ping service, I think you may miss some changes in my PR. In my test I do see it searches for the correct ping service name based on parameters defined, for example, here is one from my previous test:

2019-06-30 15:03:58,122 INFO  [org.openshift.ping.common.Utils] 3 attempt(s) with a 1000ms sleep to execute [GetServiceHosts] failed. Last failure was [java.net.UnknownHostException: my-broker-app-myping]

Having said that I may have ignored some configuration that can cause this problem. I'll keep investigating.

Howard

gaohoward commented:
@RoddieKieley  Hi Roddie,  I've found one issue with my PR. I didn't added the new environment variables into the template and that's why the 'ping' service is not found. Now I've updated my branch with the new fix. Can you try it on your environment?
In my test with plain kubernetes I manually added those vars to my yaml so I didn't see the issue before.
Sorry about that.

Howard

howardgao committed:
ENTMQBR-1648 Don't hardcode openshift.DNS_PING service name

Now the service name takes the form of
${APPLICATION_NAME}-${PING_SVC_NAME}

If both APPLICATION_NAME and PING_SVC_NAME are absent the
DNS_PING service name will be default to 'ping' (that's the
original value). If only APPLICATION_NAME is absent, the service
name is the value of PING_SVC_NAME (without prefixed with
the dash).

RoddieKieley commented:
@gaohoward This worked out of the box for me this time around, will merge!
